File: kiji-express/src/main/java/com/wibidata/lang/KijiInputFormat.java
import org.apache.hadoop.mapred.InputFormat;
import org.apache.hadoop.mapred.JobConf;
import org.apache.hadoop.mapred.RecordReader;

/**
 * Input format for Kiji that uses the old 'mapred' api.
 */
public class KijiInputFormat
    implements InputFormat<KijiKey, KijiValue> {

  /** {@inheritDoc} */
  @Override
  public InputSplit[] getSplits(JobConf conf, int numSplits) throws IOException {
    final KijiURI inputTableURI =
        KijiURI.newBuilder(conf.get(KijiConfKeys.KIJI_INPUT_TABLE_URI)).build();
    final Kiji kiji = Kiji.Factory.open(inputTableURI, conf);
    try {
      final KijiTable table = kiji.openTable(inputTableURI.getTable());
      try {
        final HTableInterface htable = HBaseKijiTable.downcast(table).getHTable();

        final List<InputSplit> splits = Lists.newArrayList();
        for (KijiRegion region : table.getRegions()) {
          final byte[] startKey = region.getStartKey();
          // TODO(KIJIMR-65): For now pick the first available location (ie. region server), if any.
          final String location =
              region.getLocations().isEmpty() ? null : region.getLocations().iterator().next();
          final TableSplit tableSplit =
              new TableSplit(htable.getTableName(), startKey, region.getEndKey(), location);
          splits.add(new KijiTableSplit(tableSplit, startKey));
        }
        return splits;

      } finally {
        table.close();
      }
    } finally {
      kiji.release();
    }
  }

  /** {@inheritDoc} */
  @Override
  public RecordReader<KijiKey, KijiValue> createRecordReader(
      InputSplit split, JobConf conf, Reporter reporter) throws IOException {
    // TODO: Use reporter to report progress.
    return new KijiTableRecordReader(split, conf);
  }

  /** Hadoop record reader for Kiji table rows. */
  public static class KijiTableRecordReader
      implements RecordReader<KijiKey, KijiValue> {

    /** Data request. */
    protected final KijiDataRequest mDataRequest;

    private Kiji mKiji = null;
    private KijiTable mTable = null;
    private KijiTableReader mReader = null;
    private KijiRowScanner mScanner = null;
    private Iterator<KijiRowData> mIterator = null;

    private KijiTableSplit mSplit = null;

    private HBaseKijiRowData mCurrentRow = null;

    /**
     * Creates a new RecordReader for this input format.
     *
     * Perform the actual reads from Kiji.
     *
     * @param conf Configuration for the target Kiji.
     */
    public KijiTableRecordReader(InputSplit split, Configuration conf) {
      // Get data request from the job configuration.
      final String dataRequestB64 = conf.get(KijiConfKeys.KIJI_INPUT_DATA_REQUEST);
      Preconditions.checkNotNull(dataRequestB64, "Missing data request in job configuration.");
      final byte[] dataRequestBytes = Base64.decodeBase64(Bytes.toBytes(dataRequestB64));
      mDataRequest = (KijiDataRequest) SerializationUtils.deserialize(dataRequestBytes);

      // Open connections to Kiji.
      assert split instanceof KijiTableSplit;
      mSplit = (KijiTableSplit) split;

      final KijiURI inputURI =
          KijiURI.newBuilder(conf.get(KijiConfKeys.KIJI_INPUT_TABLE_URI)).build();
      final KijiScannerOptions scannerOptions = new KijiScannerOptions()
          .setStartRow(new HBaseEntityId(mSplit.getStartRow()))
          .setStopRow(new HBaseEntityId(mSplit.getEndRow()));
      mKiji = Kiji.Factory.open(inputURI, conf);
      mTable = mKiji.openTable(inputURI.getTable());
      mReader = mTable.openTableReader();
      mScanner = mReader.getScanner(mDataRequest, scannerOptions);
      mIterator = mScanner.iterator();
    }

    /** {@inheritDoc} */
    @Override
    public KijiKey createKey() {
      return new KijiKey();
    }

    /** {@inheritDoc} */
    @Override
    public KijiValue createValue() {
      return new KijiValue();
    }

    /** {@inheritDoc} */
    @Override
    public long getPos() {
      // TODO: Implement
      return 0L;
    }

    /** {@inheritDoc} */
    @Override
    public float getProgress() {
      // TODO: Implement
      return 0.0f;
    }

    /** {@inheritDoc} */
    @Override
    public boolean next(KijiKey key, KijiValue value) {
      if (mIterator.hasNext()) {
        // Read the next row and store it in the provided key/value pair.
        final KijiRowData row = mIterator.next();
        key.set(row.getEntityId());
        value.set(row);

        return true;
      } else {
        return false;
      }
    }

    /** {@inheritDoc} */
    @Override
    public void close() throws IOException {
      ResourceUtils.closeOrLog(mScanner);
      ResourceUtils.closeOrLog(mReader);
      ResourceUtils.closeOrLog(mTable);
      mKiji.release();
      mIterator = null;
      mScanner = null;
      mReader = null;
      mTable = null;
      mKiji = null;

      mSplit = null;
      mCurrentRow = null;
    }
  }

  // TODO: Get rid of this.
  public static class KijiKey {
    private EntityId mCurrentKey;

    public EntityId get() {
      return mCurrentKey;
    }

    public void set(EntityId key) {
      mCurrentKey = key;
    }
  }

  // TODO: Get rid of this.
  public static class KijiValue {
    private KijiRowData mCurrentValue;

    public KijiRowData get() {
      return mCurrentValue;
    }

    public void set(KijiRowData value) {
      mCurrentValue = value;
    }
  }
}

File: kiji-express/src/main/scala/org/kiji/express/flow/framework/KijiInputFormat.scala
/**
 * (c) Copyright 2013 WibiData, Inc.
 *
 * See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kiji.express.flow.framework

import scala.collection.JavaConverters._

import com.google.common.base.Preconditions.checkNotNull
import org.apache.hadoop.hbase.mapreduce.TableSplit
import org.apache.hadoop.mapred.InputFormat
import org.apache.hadoop.mapred.InputSplit
import org.apache.hadoop.mapred.JobConf
import org.apache.hadoop.mapred.RecordReader
import org.apache.hadoop.mapred.Reporter

import org.kiji.annotations.ApiAudience
import org.kiji.annotations.ApiStability
import org.kiji.annotations.Inheritance
import org.kiji.express.util.Resources.doAndClose
import org.kiji.express.util.Resources.doAndRelease
import org.kiji.mapreduce.framework.KijiConfKeys
import org.kiji.schema.Kiji
import org.kiji.schema.KijiRegion
import org.kiji.schema.KijiTable
import org.kiji.schema.KijiURI
import org.kiji.schema.impl.HBaseKijiTable

/**
 * Tells the MapReduce framework how to divide a Kiji table into input splits for tasks,
 * and how to read records from those splits.
 *
 * MapReduce views a data set as a collection of key-value pairs divided into input splits,
 * where each input split is processed by a MapReduce task. This input format divides a Kiji
 * table into one input split per HBase region in the table. It also provides access to a record
 * reader (specifically [[org.kiji.express.flow.framework.KijiRecordReader]]) which knows how to
 * read rows from a Kiji table as key-value pairs.
 *
 * A MapReduce job reading from a Kiji table as part of the KijiExpress framework should be
 * configured with this input format. The job using this input format should have a configuration
 * containing a serialized `KijiDataRequest` at the key `kiji.input.data.request` and a Kiji URI
 * addressing the target table at the key `kiji.input.table.uri`.
 *
 * The Kiji framework already has an input format for reading from Kiji tables,
 * but it is written against a newer MapReduce API than the one supported by Cascading. This
 * input format exists to address this compatibility issue.
 */
@ApiAudience.Framework
@ApiStability.Experimental
@Inheritance.Sealed
final class KijiInputFormat
    extends InputFormat[KijiKey, KijiValue] {
  /**
   * Creates one input split per HBase region of a Kiji table.
   *
   * @param configuration containing a Kiji URI at the key `kiji.input.table.uri` that addresses
   *     the table to generate splits for.
   * @param numSplits is a MapReduce framework hint for the number of splits to produce,
   *     and is ignored here.
   * @return one input split per HBase region of the Kiji table.
   */
  override def getSplits(configuration: JobConf, numSplits: Int): Array[InputSplit] = {
    val uriString: String = checkNotNull(configuration.get(KijiConfKeys.KIJI_INPUT_TABLE_URI))
    val inputTableURI: KijiURI = KijiURI.newBuilder(uriString).build()

    doAndRelease(Kiji.Factory.open(inputTableURI, configuration)) { kiji: Kiji =>
      doAndRelease(kiji.openTable(inputTableURI.getTable())) { table: KijiTable =>
        doAndClose(HBaseKijiTable.downcast(table).openHTableConnection()) { htable =>
          table.getRegions().asScala.map { region: KijiRegion =>
            val startKey: Array[Byte] = region.getStartKey()
            // TODO(KIJIMR-65): For now pick the first available location (ie. region server),
            //     if any.
            val location: String = {
              if (region.getLocations().isEmpty()) {
                // scalastyle:off null
                null
                // scalastyle:on null
              } else {
                region.getLocations().iterator().next()
              }
            }
            val tableSplit: TableSplit = {
              new TableSplit(
                  htable.getTableName(),
                  startKey,
                  region.getEndKey(),
                  location)
            }
            new KijiTableSplit(tableSplit)
          }
          .toArray
        }
      }
    }
  }

  /**
   * Creates a record reader that will read rows from a region of a Kiji table as key-value pairs.
   *
   * @param split identifies the HBase region of the Kiji table that should be read.
   * @param configuration containing a serialized data request at the key
   *     `kiji.input.data.request` which will be used to configure how data will be read from
   *     HBase.
   * @param reporter is provided by the MapReduce framework as a means for tasks to report status,
   *     and is ignored here.
   * @return a record reader that will read rows from a region of a Kiji table as key-value pairs.
   */
  override def getRecordReader(
      split: InputSplit,
      configuration: JobConf,
      reporter: Reporter): RecordReader[KijiKey, KijiValue] = {
    split match {
      // TODO: Use reporter to report progress.
      case kijiSplit: KijiTableSplit => new KijiRecordReader(kijiSplit, configuration)
      case _ => sys.error("KijiInputFormat requires a KijiTableSplit.")
    }
  }
}
